lock-api/src/main/java/com/palantir/lock/watch/ClientLockWatchEventLogImpl.java

```diff
@@ -0,0 +1,194 @@
+        this.snapshotUpdater = snapshotUpdater;
+    }
+
+    @Override
+    public synchronized Optional<IdentifiedVersion> processUpdate(LockWatchStateUpdate update) {
+        ensureNotFailed(() -> {
+            final ProcessingVisitor visitor;
+            if (!latestVersion.isPresent() || !update.logId().equals(latestVersion.get().id())) {
+                visitor = new NewLeaderVisitor();
+            } else {
+                visitor = new ProcessingVisitor();
+            }
+            update.accept(visitor);
+        });
+        return latestVersion;
+    }
+
+    @Override
+    public synchronized void removeOldEntries(IdentifiedVersion earliestVersion) {
+        ensureNotFailed(() -> {
+            Set<Map.Entry<Long, LockWatchEvent>> eventsToBeRemoved =
+                    eventMap.headMap(earliestVersion.version()).entrySet();
+            snapshotUpdater.processEvents(
+                    eventsToBeRemoved.stream().map(Map.Entry::getValue).collect(Collectors.toList()));
+            eventsToBeRemoved.clear();
+        });
+    }
+
+    /**
+     * @param startVersion latest version that the client knows about; should be before timestamps in the mapping;
+     * @param endVersion   mapping from timestamp to identified version from client-side event cache;
+     * @return lock watch events that occurred from (exclusive) the provided version, up to (inclusive) the latest
+     * version in the timestamp to version map.
+     */
+    @Override
+    public synchronized ImmutableTransactionsLockWatchEvents.Builder getEventsBetweenVersions(
+            Optional<IdentifiedVersion> startVersion,
+            IdentifiedVersion endVersion) {
+        checkNotFailed();
+        Optional<IdentifiedVersion> versionInclusive = startVersion.map(this::createInclusiveVersion);
+        IdentifiedVersion currentVersion = getLatestVersionAndVerify(endVersion);
+        ImmutableTransactionsLockWatchEvents.Builder eventBuilder = ImmutableTransactionsLockWatchEvents.builder();
+        List<LockWatchEvent> events = new ArrayList<>();
+        final long fromSequence;
+
+        if (!versionInclusive.isPresent() || differentLeaderOrTooFarBehind(currentVersion, versionInclusive.get())) {
+            events.add(LockWatchCreatedEvent.fromSnapshot(snapshotUpdater.getSnapshot(currentVersion)));
+            fromSequence = eventMap.firstKey();
+            eventBuilder.clearCache(true);
+        } else {
+            fromSequence = versionInclusive.get().version();
+            eventBuilder.clearCache(false);
+        }
+
+        events.addAll(eventMap.subMap(fromSequence, INCLUSIVE, endVersion.version(), INCLUSIVE).values());
+        return eventBuilder.events(events);
+    }
+
+    @Override
+    public synchronized Optional<IdentifiedVersion> getLatestKnownVersion() {
+        checkNotFailed();
+        return latestVersion;
+    }
+
+    private synchronized void ensureNotFailed(Runnable runnable) {
+        checkNotFailed();
```

Comment:
So I understand that you added this in response to @j-baker comment, but I think as it is it's not really adding much value, beyond detecting a failure and not letting yourself continue to get corrupted. I quite like the defensiveness here in case we have bugs, but I think you need to finish the stroke and switch to a NoOpCache at runtime.

Also, I would argue this will be implemented cleaner as FailureCheckingLockWatchEventCache, as a proxy.

Reply:
Done